src/checkers/entropy.js

```diff
@@ -0,0 +1,180 @@
+          )}
+        </>
+      );
+  };
+
+const determineStrength = (password) =>{
+    const length = password.length;
+    var bonus = 0;
+    var i = 0;
+    var uppercase = false;
+    var lowercase = false;
+    var special = false;
+    var number = false;
+    var charsetSize = 0;
+    var entropy = 0;
+    //Bonus bits-------------------
+    /*
+    //Length bonuses
+    if(length >= 1){    //First character. +4
+        bonus += 4.0;
+        if(length >= 2){  //2nd-8th characters. +2
+            if(length > 8){
+                bonus += 7 * 2;
+                if(length > 20){    //9th-20th characters. +1.5
+                    bonus += 1.5 * 12
+                }
+                else{
+                    bonus += (length - 8) * 1.5;
+                }
+            }
+            else{
+                bonus += (length-1) * 2.0;
+            }
+        }
+    }
+*/
+    //Contains uppercase or non-alphabetic characters. bonus +3 for both
+    //Other conditionals are for getting the charset size
+    while(i < length){
+        var char = password[i];
+        if(isNaN(char * 1)){   //turns from char to int, checks if not a number
+            if(char.match(/[A-Z]/g)){    //check if uppercase
+                uppercase = true;
+            }
+            else if(char.match(/[a-z]/g)){   //check if lowercase
+                lowercase = true;
+            }
+            else{ //Else it's not alphanumeric
+                special = true;
+            }
+        }
+        else{
+            number = true;
+        }
+        i++;
+    }
+
+    //Dictionary check. +3 if substring matches, +6 if no match
+    //TODO: ^^
+
+    //get charset size
+    if(uppercase){charsetSize += 26;}
+    if(lowercase){charsetSize += 26;}
+    if(special){charsetSize += 33;}
+    if(number){charsetSize += 10;}
+
+    //entropy formula according to NIST Eletronic Authentication Guidelines
+    if(charsetSize > 0){
+        entropy = Math.log2(charsetSize) * length;
+    } else{ 
+        entropy = 0;
+    }
+
+    console.log("Entropy: " + entropy + "Bonus bits: " + bonus);
+
+    var feedback = {
+        score: entropy,
+        uppercase: uppercase,
+        lowercase: lowercase,
+        special: special,
+        number: number,
+    };
+
+    return generateFeedback(feedback);
+};
+
+
+
+const generateFeedback = (result) => {
+    var score = 0;
+    var description = [];
+    var feedback = {};
+    
+    //Charset suggestions
+    if(!result.uppercase){ description.push("Use uppercase letters");}
+    if(!result.lowercase){description.push("Use lowercase letters");}
+    if(!result.special){description.push("Use special symbols");}
+    if(!result.number){description.push("Use numbers");}
+
+    //Scoring sorta from this website: https://www.pleacher.com/mp/mlessons/algebra/entropy.html
+    //<14 = 1
+    //<28 = very weak = 2-3
+    //28-35 weak = 4-5
+    //36-59 reasonable  = 6-7
+    //60-127 strong = 8-9
+    //128+ very strong = 10
+
```

Comment:
Using these scores, "password" comes out as a strength of 6, which didn't really make sense so I tried to make up my own thresholds below. If we include bonus bits, these will need to be raised quite a bit, so it's undecided if those will be used

Reply:
FWIW if we followed the algorithm correctly and it's doing stuff like ranking "password" as a 6 that's actually fine, that helps show the strengths/weaknesses of the different approaches